File: bindings/scala/src/main/scala/dev/everanium/itb/Opts.scala
// Immutable options builder mirroring the Java binding's Opts.
//
// The builder performs no validation — every key and value is passed
// through to Go verbatim; libitb rejects unknown keys or bad values
// with a diagnostic surfaced via ItbError. Primitive / MAC / cipher /
// palette names are opaque strings.

package dev.everanium.itb

import com.everanium.itb.Opts as JOpts

/** Immutable, chainable options for [[Pipeline.init]],
  * [[Pipeline.open]], and [[Pipeline.registerProfile]]. An empty
  * value renders the empty query (pure profile defaults). Each
  * `with*` call returns a new value; sharing a prefix between two
  * configurations is safe.
  */
final case class Opts(pairs: Vector[(String, String)] = Vector.empty):

  /** Escape hatch appending a raw `key=value` pair. Covers every key
    * the Go side accepts, including the register-profile grammar
    * (`mode`, `width`, `innerHashes`, `parallaxOn`, `wrapperOn`, …).
    */
  def withRaw(key: String, value: String): Opts = copy(pairs :+ (key -> value))

  /** Hex-encodes the parallax master override (`pm`). */
  def withPermMaster(master: Array[Byte]): Opts = withRaw("pm", Opts.hex(master))

  /** Hex-encodes the wrapper master override (`wm`). */
  def withWrapMaster(master: Array[Byte]): Opts = withRaw("wm", Opts.hex(master))

  def withParallax(on: Boolean): Opts = withRaw("withParallax", on.toString)

  def withWrapper(on: Boolean): Opts = withRaw("withWrapper", on.toString)

  def withMaxWorkers(n: Long): Opts = withRaw("maxWorkers", n.toString)

  def withNonceBits(n: Long): Opts = withRaw("nonceBits", n.toString)

  def withBarrierFill(n: Long): Opts = withRaw("barrierFill", n.toString)

  def withChunkSize(n: Long): Opts = withRaw("chunkSize", n.toString)

  def withKeyBits(n: Long): Opts = withRaw("keyBits", n.toString)

  def withParallaxSegmentSize(n: Long): Opts = withRaw("parallaxSegmentSize", n.toString)

  def withMacName(name: String): Opts = withRaw("macName", name)

  def withInnerHash(name: String): Opts = withRaw("innerHash", name)

  def withOuterCipher(name: String): Opts = withRaw("outerCipher", name)

  /** Comma-joins the palette names (`parallaxPalette`). */
  def withParallaxPalette(names: String*): Opts =
    withRaw("parallaxPalette", names.mkString(","))

  /** Replays the accumulated pairs into a fresh Java builder; the
    * Java side owns query rendering and percent-encoding.
    */
  private[itb] def toJava: JOpts =
    pairs.foldLeft(new JOpts()) { case (o, (k, v)) => o.withRaw(k, v) }

object Opts:
  /** The empty options value (pure profile defaults). */
  val empty: Opts = Opts()

  private def hex(bytes: Array[Byte]): String =
    val sb = new StringBuilder(bytes.length * 2)
    bytes.foreach(b => sb.append(f"${b & 0xff}%02x"))
    sb.toString

File: bindings/vbnet/src/EveraniumItb.VisualBasic/Opts.vb
' Chaining wrapper over the C# layer's URL-query opts builder.
'
' The builder performs no validation — every key and value is
' rendered into a percent-encoded query string and passed through to
' Go verbatim; libitb rejects unknown keys or bad values with a
' diagnostic surfaced via ItbException. Primitive / MAC / cipher /
' palette names are opaque strings.

''' <summary>
''' Builder producing the opts string consumed by
''' <see cref="Pipeline.Init"/>, <see cref="Pipeline.Open"/>, and
''' <see cref="Pipeline.RegisterProfile"/>. Setters chain; an empty
''' builder renders the empty query (pure profile defaults).
''' </summary>
Public NotInheritable Class Opts

    Private ReadOnly _inner As New Global.Itb.Opts()

    ''' <summary>The wrapped C# builder handed to the inner layer's
    ''' entry points.</summary>
    Friend ReadOnly Property Inner As Global.Itb.Opts
        Get
            Return _inner
        End Get
    End Property

    ''' <summary>Hex-encodes the parallax master override
    ''' (<c>pm</c>).</summary>
    Public Function WithPermMaster(master As Byte()) As Opts
        _inner.WithPermMaster(master)
        Return Me
    End Function

    ''' <summary>Hex-encodes the wrapper master override
    ''' (<c>wm</c>).</summary>
    Public Function WithWrapMaster(master As Byte()) As Opts
        _inner.WithWrapMaster(master)
        Return Me
    End Function

    Public Function WithParallax([on] As Boolean) As Opts
        _inner.WithParallax([on])
        Return Me
    End Function

    Public Function WithWrapper([on] As Boolean) As Opts
        _inner.WithWrapper([on])
        Return Me
    End Function

    Public Function WithMaxWorkers(n As Long) As Opts
        _inner.WithMaxWorkers(n)
        Return Me
    End Function

    Public Function WithNonceBits(n As Long) As Opts
        _inner.WithNonceBits(n)
        Return Me
    End Function

    Public Function WithBarrierFill(n As Long) As Opts
        _inner.WithBarrierFill(n)
        Return Me
    End Function

    Public Function WithChunkSize(n As Long) As Opts
        _inner.WithChunkSize(n)
        Return Me
    End Function

    Public Function WithKeyBits(n As Long) As Opts
        _inner.WithKeyBits(n)
        Return Me
    End Function

    Public Function WithParallaxSegmentSize(n As Long) As Opts
        _inner.WithParallaxSegmentSize(n)
        Return Me
    End Function

    Public Function WithMacName(name As String) As Opts
        _inner.WithMacName(name)
        Return Me
    End Function

    Public Function WithInnerHash(name As String) As Opts
        _inner.WithInnerHash(name)
        Return Me
    End Function

    ''' <summary>Comma-joins an 8-slot per-call inner-hash
    ''' constellation into the <c>innerHashes</c> opts key. Parallel
    ''' to the Go-side <c>Opts.MixedHashes [8]string</c> per-call
    ''' override; slot ordering is
    ''' <c>[noise, lock, data1, data2, data3, start1, start2, start3]</c>.
    ''' Fail-fast validation surfaces at Init on the Go side; a
    ''' typo'd slot or width mismatch surfaces with an error naming
    ''' the offending slot. When both this and
    ''' <see cref="WithInnerHash"/> are set, the mixed override wins
    ''' on the Go side.</summary>
    Public Function WithInnerHashes(ParamArray names As String()) As Opts
        _inner.WithInnerHashes(names)
        Return Me
    End Function

    Public Function WithOuterCipher(name As String) As Opts
        _inner.WithOuterCipher(name)
        Return Me
    End Function

    ''' <summary>Comma-joins the palette names
    ''' (<c>parallaxPalette</c>).</summary>
    Public Function WithParallaxPalette(ParamArray names As String()) As Opts
        _inner.WithParallaxPalette(names)
        Return Me
    End Function

    ''' <summary>Escape hatch appending a raw <c>key=value</c> pair.
    ''' Covers every key the Go side accepts, including the
    ''' register-profile grammar (<c>mode</c>, <c>width</c>,
    ''' <c>innerHashes</c>, <c>parallaxOn</c>, <c>wrapperOn</c>, …).
    ''' </summary>
    Public Function WithRaw(key As String, value As String) As Opts
        _inner.WithRaw(key, value)
        Return Me
    End Function
End Class
